StyleCop.Analyzers/StyleCop.Analyzers/GeneratedCodeAnalysisExtensions.cs

```diff
@@ -0,0 +1,97 @@
+
+        internal static bool IsGenerated(this SymbolAnalysisContext context, CancellationToken cancellationToken)
+        {
+            if (context.Symbol == null) return false;
+            foreach (var syntaxReference in context.Symbol.DeclaringSyntaxReferences)
+            {
+                if (syntaxReference.SyntaxTree.IsGenerated(cancellationToken)) return true;
+                var root = syntaxReference.SyntaxTree.GetRoot(cancellationToken);
+                var node = root?.FindNode(syntaxReference.Span);
+                if (node.IsGenerated()) return true;
+            }
+            return false;
+        }
+
+        internal static bool IsGenerated(this SyntaxNode node) => node.HasAttributeOnAncestorOrSelf("DebuggerNonUserCode", "GeneratedCode");
+
+        internal static bool IsGenerated(this SyntaxTree tree, CancellationToken cancellationToken) => (tree.FilePath?.IsOnGeneratedFile() ?? false) || tree.HasAutoGeneratedComment(cancellationToken);
+
+        internal static bool HasAutoGeneratedComment(this SyntaxTree tree, CancellationToken cancellationToken)
+        {
+            StrongBox<bool?> cachedResult = GeneratedHeaderPresentCheck.GetOrCreateValue(tree);
+            if (cachedResult.Value.HasValue)
+            {
+                return cachedResult.Value.Value;
+            }
+
+            bool autoGenerated = HasAutoGeneratedCommentNoCache(tree, cancellationToken);
+
+            // Update the strongbox's value with our computed result.
+            // This doesn't change the strongbox reference, and its presence in the
+            // ConditionalWeakTable is already assured, so we're updating in-place.
+            // In the event of a race condition with another thread that set the value,
+            // we'll just be re-setting it to the same value.
+            cachedResult.Value = autoGenerated;
+
+            return autoGenerated;
+        }
+
+        private static bool HasAutoGeneratedCommentNoCache(SyntaxTree tree, CancellationToken cancellationToken)
+        {
+            var root = tree.GetRoot(cancellationToken);
+
+            if (root == null) return false;
+            var firstToken = root.GetFirstToken();
+            SyntaxTriviaList trivia;
+            if (firstToken == default(SyntaxToken))
+            {
+                var token = ((CompilationUnitSyntax)root).EndOfFileToken;
+                if (!token.HasLeadingTrivia) return false;
+                trivia = token.LeadingTrivia;
+            }
+            else
+            {
+                if (!firstToken.HasLeadingTrivia) return false;
+                trivia = firstToken.LeadingTrivia;
+            }
+
+            var commentLines = trivia.Where(t => t.IsKind(SyntaxKind.SingleLineCommentTrivia)).Take(2).ToList();
+            if (commentLines.Count != 2) return false;
+            return commentLines[1].ToString() == "// <auto-generated>";
```

Comment:
This may be "too strict", in the sense that it fails to exclude files that people assume would be excluded (e.g. due to spacing differences). Of particular concern is the example of generated code given in the [documentation for SA1633](http://stylecop.com/docs/SA1633.html) would not be considered auto-generated according to this implementation (both for being on the first line, and for being a self-closing XML element instead of the one shown here).